neo/Storage/Implementation/AbstractDBImplementation.py

```diff
@@ -0,0 +1,204 @@
+    Alternative backends can be added by implementing the required methods
+    of this abstract class.
+Usage:
+    The dynamically generated class coming from the database factory inherits
+    from the abstract class, means the generated class cannot be used if not
+    all methods defined in this class are overwritten.
+
+    If this class is extended, make sure you extend also all other database
+    implementations.
+
+    For a more detailed description on the methods and how to implement a new
+    database backend check out:
+    neo.Storage.Implementation.LevelDB.LevelDBImpl
+"""
+
+
+class AbstractDBImplementation(ABC):
+
+    @abstractmethod
+    def __init__(self, path):
+        """
+        Init method used within the DBFactory, opens a new or existing database.
+
+        Args:
+            path (str): full path to the database directory.
+
+        Attributes:
+            path (str): full path to the database directory.
+            _db (object): the database instance
+
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def write(self, key, value):
+        """
+        Writes the given key/value pair to the database.
+
+        Args:
+            key (bytearray): has to be a prefixed bytearray, for prefixes check
+                             neo.Storage.Common.DBPrefix
+
+            value (bytearray): the value as bytearray
+
+        Returns:
+            None
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, key):
+        """
+        Retrieves the value based on the given key from the database.
+
+        Args:
+            key (bytearray): has to be a prefixed bytearray, for prefixes check
+                             neo.Storage.Common.DBPrefix
+
+        Returns:
+            bytearray
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def delete(self, key):
+        """
+        Deletes a row from the database.
+
+        Args:
+            key (bytearray): has to be a prefixed bytearray, for prefixes check
+                             neo.Storage.Common.DBPrefix
+
+        Returns:
+            None
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def cloneDatabase(self, clone_db):
+        """
+        Clones the current database into "clone_db"
+
+        Args:
+            clone_db (object): the instance of the database to clone to.
+
+        Returns:
+            clone_db (object): returns a cloned db instance
+
+        """
+
+        raise NotImplementedError
+
+    @abstractmethod
+    def createSnapshot(self):
+        """
+        Creates a snapshot of the current database, used for DebugStorage and
+        NotificationDB. To keep the snapshot compatible to the current design it's
+        created through a factory which returns basically the same class we use
+        for the real database and all the methods that can be used on the real db
+        can also be used on the snapshot.
```

Comment:
I think we can remove this part
```
       To keep the snapshot compatible to the current design it's
        created through a factory which returns basically the same class we use
        for the real database and all the methods that can be used on the real db
        can also be used on the snapshot.
```
as this looks like implementation details and an explanation towards me why we 'used' a factory (which no longer applies).

Also **snapshot** to me means a copy of state at a particular point in time. Now the discussion point would be; should it be read-only or not? 
[Leveldb's](https://github.com/google/leveldb/blob/master/doc/index.md#snapshots) snapshot feature is readonly,  SQLite also seems to be [readonly](https://www.sqlite.org/c3ref/snapshot.html) although they don't describe it as implicit as leveldb. I'm currently leaning towards having read-only as well, which currently would only mean updating the documentation to ensure that it reads 
```
Creates a read-only snapshot of the current database, used for DebugStorage and NotificationDB
```

Reply:
Done.
> Now the discussion point would be; should it be read-only or not?
I think it has to be read-only, at least for now, with `LevelDB` as the default database there is no other option. Aside from that, if you have multiple snapshots changing the same data, you can run into serious problems if there is no transaction handling. I think this is something to discuss for the future, maybe keeping it read-only is sufficient?